src/debug/goDebugProvider.ts

```diff
@@ -0,0 +1,123 @@
+import * as debugUtils from "./debugUtils";
+import * as extensionUtils from "../extensionUtils";
+import { Kubectl } from "../kubectl";
+import { IDockerfile } from "../docker/parser";
+
+const defaultGoDebuggerExtensionId = "golang.go";
+const defaultGoDebuggerExtension = "Debugger for Go";
+const defaultGoDebuggerConfigType = "go";
+
+const defaultGoAppPort = "8080";
+const defaultGoDebugPort = "32768";
+const fullGoDebugOptsRegExp = /^dlv debug\s+.*(--listen[\s=][^\s]+)\S*/i;
+
+export class GoDebugProvider implements IDebugProvider {
+    public getDebuggerType(): string {
+        return defaultGoDebuggerConfigType;
+    }
+
+    public async isDebuggerInstalled(): Promise<boolean> {
+        if (vscode.extensions.getExtension(defaultGoDebuggerExtensionId)) {
+            return true;
+        }
+        const answer = await vscode.window.showInformationMessage(`Go debugging requires the '${defaultGoDebuggerExtension}' extension. Would you like to install it now?`, "Install Now");
+        if (answer === "Install Now") {
+            return await extensionUtils.installVscodeExtension(defaultGoDebuggerExtensionId);
+        }
+        return false;
+    }
+
+    public async startDebugging(workspaceFolder: string, sessionName: string, port: number | undefined, _pod: string, _pidToDebug: number | undefined): Promise<boolean> {
+        const debugConfiguration = {
+            type: "go",
+            request: "attach",
+            mode: "remote",
+            name: sessionName,
+            hostName: "localhost",
+            remotePath: "${inputs.remotePath}",
+            port,
+            inputs: [
+                {
+                    id: "remotePath",
+                    type: "promptString",
+                    description: "What is the remote dirpath for your app?",
+                    default: "/go/src/${workspaceFolderBasename}"
+                }
+            ]
+        };
+        const currentFolder = (vscode.workspace.workspaceFolders || []).find((folder) => folder.name === path.basename(workspaceFolder));
+        if (!currentFolder) {
+            return false;  // shouldn't happen as workspaceFolder should be one of the workspaceFolders
+        }
+        return await vscode.debug.startDebugging(currentFolder, debugConfiguration);
+    }
+
+    public isSupportedImage(baseImage: string): boolean {
+        if (!baseImage) {
+            return false;
+        }
+        return baseImage.indexOf("golang") >= 0;
+    }
+
+    public async resolvePortsFromFile(dockerfile: IDockerfile, env: {}): Promise<PortInfo | undefined> {
+        // Resolve the debug port.
+        const rawDebugPortInfo = await debugUtils.promptForDebugPort(defaultGoDebugPort);
+
+        // Resolve the app port.
+        const exposedPorts = dockerfile.getExposedPorts();
+        const possiblePorts = exposedPorts.length ? exposedPorts.filter((port) => port !== defaultGoDebugPort) : [];
```

Comment:
Should we also exclude the actual debug port if different from the default one?

Reply:
That's a good point. Especially since they just were prompted for that port, I think we can exclude that one from the list.